solutions/usaco-813.mdx

```diff
@@ -5,54 +5,248 @@ title: Snow Boots
 author: Benjamin Qi
 ---
 
-[Official Analysis](http://www.usaco.org/current/data/sol_snowboots_gold_feb18.html)
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_snowboots_gold_feb18.html)
 
 We can use a sorted set in place of the linked list mentioned in the official
-solution. The time complexity remains the same.
+solution.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N \log N + B \log B)$
+
+<LanguageSection>
+<CPPSection>
+
+We can use a sorted set in place of the linked list mentioned in the official
+solution.
 
 ```cpp
 #include <bits/stdc++.h>
 using namespace std;
 
 int main() {
-	freopen("snowboots.in","r",stdin);
-	freopen("snowboots.out","w",stdout);
+	freopen("snowboots.in", "r", stdin);
+
+	int tile_num;
+	int boot_num;
+	cin >> tile_num >> boot_num;
 
-	int N,B; cin >> N >> B;
-	vector<int> f(N); for (int& t: f) cin >> t;
+	vector<int> tiles(tile_num);
+	for (int& t: tiles) {
+		cin >> t;
+	}
+
+	struct Boot { int max_depth, max_steps, index; };
+	vector<Boot> boots(boot_num);
+	for (int i = 0; i < boot_num; ++i) {
+		cin >> boots[i].max_depth >> boots[i].max_steps;
+		boots[i].index = i;
+	}
+	// sort the boots from greatest depth to least depth
+	sort(
+		boots.begin(), boots.end(),
+		[&](const Boot& a, const Boot& b) { return a.max_depth > b.max_depth; }
+	);
 
-	vector<int> tiles_by_depth;
-	for (int i = 1; i < N-1; ++i)
+	vector<int> tiles_by_depth;  // list of tile indices, sorted by snow depth
+	for (int i = 1; i < tile_num - 1; i++) {
 		tiles_by_depth.push_back(i);
-	sort(begin(tiles_by_depth),end(tiles_by_depth),
-		[&](int x, int y) { return f[x] > f[y]; });
+	}
+	sort(
+		tiles_by_depth.begin(), tiles_by_depth.end(),
+		[&](int a, int b) { return tiles[a] > tiles[b]; }
+	);
+
+	set<int> valid_tiles;
+	for (int i = 0; i < tile_num; i++) {
+		valid_tiles.insert(i);
+	}
 
-	set<int> indices;
-	for (int i = 0; i < N; ++i)
-		indices.insert(i);
+	// the deepest tile that can still be traversed with the boot we're at
+	int tile_at = 0;
```

Comment:
`till_at` might cause confusion as it does not describe the *position* of the tile from the input. It is instead used to iterate through the sorted tiles. Maybe `removed_tiles_count`?

Reply:
yeah, it's the index of the *tile* we're _at_

Reply:
Idk.. But I think it's understandable so it's alright